File: api/functions/app/messages.py
"""Centralized error and status messages for localization."""

# ─── Auth ────────────────────────────────────────────────────────────────────
AUTH_REQUIRED = "Authentication required"
AUTH_MISSING_HEADER = "Missing or malformed Authorization header"
AUTH_INVALID_TOKEN = "Invalid or expired token"

# ─── Admin ───────────────────────────────────────────────────────────────────
ADMIN_INVALID_EMAIL = "Please enter a valid email address"
ADMIN_ALREADY_EXISTS = "This email is already in the allowlist"
ADMIN_NOT_FOUND = "This email is not in the allowlist"
ADMIN_CANNOT_REMOVE_SELF = "You cannot remove your own email"
ADMIN_CANNOT_REMOVE_LAST = "Cannot remove the last admin"
ADMIN_CANNOT_REVOKE_LAST = "Cannot revoke admin from the last admin"

# ─── Habits ──────────────────────────────────────────────────────────────────
HABIT_NAME_REQUIRED = "Habit name is required"
HABIT_NOT_FOUND = "Habit not found"
HABIT_INVALID_FREQUENCY = "Invalid frequency configuration"
HABIT_INVALID_DATE = "Invalid date format"
HABIT_FUTURE_DATE = "Cannot log completion for a future date"
HABIT_REMINDER_TIME_REQUIRED = "Reminder time is required when reminders are enabled"

# ─── API Keys ───────────────────────────────────────────────────────────────
APIKEY_INVALID = "Invalid API key"
APIKEY_EXPIRED = "API key has expired"
APIKEY_REVOKED = "API key has been revoked"
APIKEY_RATE_LIMITED = "Rate limit exceeded. Try again later."
APIKEY_PERMISSION_REQUIRED = "API key management permission required"
APIKEY_MAX_KEYS = "Maximum of 2 active API keys allowed"
APIKEY_NOT_FOUND = "API key not found"
APIKEY_BEARER_REQUIRED = "API key management requires Bearer token authentication"
APIKEY_NAME_REQUIRED = "API key name is required"
APIKEY_EXPIRY_REQUIRED = "API key expiration date is required"
APIKEY_EXPIRY_PAST = "Expiration date must be in the future"

# ─── Validation ──────────────────────────────────────────────────────────────
VALIDATION_INVALID_JSON = "Invalid JSON"

File: web/src/messages.ts
/**
 * Centralized UI messages for localization.
 * API error messages are mapped to friendlier UI messages here.
 */

const API_TO_UI: Record<string, string> = {
  // Admin
  'Please enter a valid email address': 'Please enter a valid email address.',
  'This email is already in the allowlist': 'This email already has access.',
  'This email is not in the allowlist': 'This email is not in the access list.',
  'You cannot remove your own email': 'You cannot remove yourself.',
  'Cannot remove the last admin': 'There must be at least one admin.',
  'Cannot revoke admin from the last admin': 'There must be at least one admin.',

  // Auth
  'Authentication required': 'Please sign in to continue.',
};

/**
 * Convert an API error message to a user-friendly UI message.
 * Falls back to a generic message if no mapping exists.
 */
export function toUIMessage(error: unknown, fallback = 'Something went wrong. Please try again.'): string {
  if (error instanceof Error) {
    // Apollo wraps GraphQL errors — extract the message
    const msg = error.message.replace('GraphQL error: ', '');
    return API_TO_UI[msg] || msg;
  }
  return fallback;
}
